### PR TITLE
Update to support Phoenix v1.7

### PR DESCRIPTION
**Pull request by elepedus**
Updates `MainProxy.Cowboy2Handler` to use `Plug.Cowboy.Handler` instead of `Phoenix.Endpoint.Cowboy2Handler`, which will be removed in Phoenix 1.7. 
[#28]

### CODE REVIEW

**Comment by elepedus** [comment]
Ok, I've just pulled this branch into my test app, and unfortunately, it looks like sockets don't work :( 

```
[error] Ranch protocol #PID<0.1123.0> of listener HTTP (connection #PID<0.1122.0>, stream id 1) terminated
an exception was raised:
    ** (FunctionClauseError) no function clause matching in Plug.Cowboy.Handler.upgrade/5
        (plug_cowboy 2.6.0) lib/plug/cowboy/handler.ex:33: Plug.Cowboy.Handler.upgrade(%{bindings: %{}, body_length: 0, cert: :undefined, has_body: false, headers: %{"accept-encoding" => "gzip, deflate", "accept-language" => "en-GB,en-US;q=0.9,en;q=0.8", "cache-control" => "no-cache", "connection" => "Upgrade", "cookie" => "_inventory_key=SFMyNTY.g3QAAAABbQAAAAtfY3NyZl90b2tlbm0AAAAYZXBWcDk2ZEdYbm9YcGpfN1FqSjJqc3Q4.K0kZcawb401xP0u_6-UzKXXfFjRFBaJmCKvGiO77kt8; _inventory_web_key=SFMyNTY.g3QAAAABbQAAAAtfY3NyZl90b2tlbm0AAAAYRFR5YkZNbVp2SjgxWE90bVVPdzVzSmNO.Lipi_opXwBN3iKXVEOfGPI5xgDUY8AGhwqz1tEzI9kU", "host" => "inventory.lvh.me:4000", "origin" => "http://inventory.lvh.me:4000", "pragma" => "no-cache", "sec-websocket-extensions" => "permessage-deflate; client_max_window_bits", "sec-websocket-key" => "NUw4Ot6yh9npQBV7ujQGkg==", "sec-websocket-version" => "13", "upgrade" => "websocket", "user-agent" => "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36"}, host: "inventory.lvh.me", host_info: :undefined, method: "GET", path: "/phoenix/live_reload/socket/websocket", path_info: :undefined, peer: {{127, 0, 0, 1}, 59639}, pid: #PID<0.1122.0>, port: 4000, qs: "vsn=2.0.0", ref: HTTP, resp_headers: %{"cache-control" => "max-age=0, private, must-revalidate"}, scheme: "http", sock: {{127, 0, 0, 1}, 4000}, streamid: 1, upgrade: {:websocket, {WebSockAdapter.CowboyAdapter, {Phoenix.LiveReloader.Socket, %{}, {%{channels: %{}, channels_inverse: %{}}, %Phoenix.Socket{assigns: %{}, channel: nil, channel_pid: nil, endpoint: InventoryWeb.Endpoint, handler: Phoenix.LiveReloader.Socket, id: nil, joined: false, join_ref: nil, private: %{}, pubsub_server: Inventory.PubSub, ref: nil, serializer: Phoenix.Socket.V2.JSONSerializer, topic: nil, transport: :websocket, transport_pid: nil}}}, %{compress: false, idle_timeout: 60000}}}, version: :"HTTP/1.1"}, %{dispatch: [{:_, [], [{:_, [], MainProxy.Cowboy2Handler, {nil, [backends: [%{host: ~r/.*/, path: ~r/\/health/, phoenix_endpoint: HealthcheckWeb.Endpoint}, %{host: ~r/^inventory\.lvh\.me$/, phoenix_endpoint: InventoryWeb.Endpoint}], callback_module: Proxy]}}]}], handler: MainProxy.Cowboy2Handler, handler_opts: {nil, [backends: [%{host: ~r/.*/, path: ~r/\/health/, phoenix_endpoint: HealthcheckWeb.Endpoint}, %{host: ~r/^inventory\.lvh\.me$/, phoenix_endpoint: InventoryWeb.Endpoint}], callback_module: Proxy]}}, MainProxy.Cowboy2Handler, {WebSockAdapter.CowboyAdapter, {Phoenix.LiveReloader.Socket, %{}, {%{channels: %{}, channels_inverse: %{}}, %Phoenix.Socket{assigns: %{}, channel: nil, channel_pid: nil, endpoint: InventoryWeb.Endpoint, handler: Phoenix.LiveReloader.Socket, id: nil, joined: false, join_ref: nil, private: %{}, pubsub_server: Inventory.PubSub, ref: nil, serializer: Phoenix.Socket.V2.JSONSerializer, topic: nil, transport: :websocket, transport_pid: nil}}}}, %{compress: false, idle_timeout: 60000})
        (cowboy 2.9.0) /app_umbrella/deps/cowboy/src/cowboy_stream_h.erl:306: :cowboy_stream_h.execute/3
        (cowboy 2.9.0) /app_umbrella/deps/cowboy/src/cowboy_stream_h.erl:295: :cowboy_stream_h.request_process/3
        (stdlib 4.1.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3

```

**Comment by Malian** [comment]
We started a new project with Phoenix 1.7 and encountered the same problem. For now, we have rollback to Phoenix 1.6. I admit that my skills are limited, but if someone can guide me, it would be a pleasure to try tackling this issue.

**Comment by mtrudel** [comment]
Possibly of interest: 

https://elixirforum.com/t/mainproxy-run-multiple-phoenix-apps-on-one-port/51067/5?u=mtrudel

tl;dr: from the perspective of MainProxy, Phoenix 1.7 apps can be considered entirely as Plug apps (with the application Endpoint being the Plug) and things should 'just work'. The main point of the 1.7+ structure is to simplify the architecture of all this, and, TBH (I mean this in the literal sense and NOT AT ALL as a diss), I'm pretty sure you're overthinking the approach here.

**Comment by axelson** [comment]
@elepedus and @Malian I've added support for Phoenix 1.7-rc.2, could you try it out and let me know how it goes? See #31 for details.

**Comment by axelson** [comment]
Closed in favor of https://github.com/Main-Proxy/main_proxy/pull/31

Thank you for taking a look!
